app/services/puppet_class_importer.rb

```diff
@@ -330,11 +330,16 @@ def find_existing_foreman_classes(klasses = [])
   end
 
   def find_or_create_env(env)
-    Environment.where(:name => env).first || Environment.create!(:name => env,
+    user_visible_environment(env) || Environment.create!(:name => env,
                                                                  :organizations => User.current.my_organizations,
                                                                  :locations => User.current.my_locations)
   end
 
+  def user_visible_environment(env)
+    return unless User.current.visible_environments.include? env
+    Environment.unscoped.find_by :name => env
```

Comment:
Is this disregarding the User permissions completely?

Reply:
This is a variation of the problem above, it errors even for admin.

1) create 2 organizations, A and B
2) import puppet env as admin in context of A - it gets imported to B as well
3) unassign the newly imported env from B
4) in context of B, try to import the env you just unassigned

It uses `unscoped` only to avoid hitting the uniqueness constraint, user does not know it re-used the existing record. On the outside it looks like it was just created.

I realize this is not ideal and does not solve all the issues with imports, but at leas it should fix the complete application crashes.

Reply:
Fair enough